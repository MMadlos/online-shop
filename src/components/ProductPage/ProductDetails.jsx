import AddCartBtn from "../Elements/AddCartBtn"
import Rating from "./Rating"
import promoImage from "/public/promo-image-tiny.png"

function ProductDetails({ productSelected }) {
	const { id, name, price, url, rate, countRates, description } = productSelected

	return (
		<div
			className="product-container"
			data-product-id={id}>
			<img
				src={url}
				alt="#"
				className="product-image"
			/>

			<div className="product-details">
				<img
					src={promoImage}
					alt=""
					className="promo-image"
				/>
				<div className="product-header-container">
					<h1 className="product-header">{name}</h1>
					<Rating
						rate={rate}
						countRates={countRates}
					/>
					<p className="price">{`$ ${price}`}</p>
				</div>

				<AddCartBtn />

				<div className="product-about-container">
					<p>About the product</p>
					<div className="product-about">
						<p>{description}</p>
					</div>
				</div>
			</div>
		</div>
	)
}

export default ProductDetails
