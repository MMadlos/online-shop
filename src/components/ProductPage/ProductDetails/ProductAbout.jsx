import { useState } from "react"

function ProductAbout({ description }) {
	const [isSeeMoreOpen, setIsSeeMoreOpen] = useState(false)

	const expandBtn = {
		text: isSeeMoreOpen ? "See less" : "See more",
		icon: isSeeMoreOpen ? "fa-chevron-up" : "fa-chevron-down",
	}

	return (
		<div className="product-about-container">
			<p>About the product</p>
			<div className="product-about">
				<p className={isSeeMoreOpen ? "description full" : "description"}>{description}</p>
				<button
					className="expand-btn none"
					onClick={() => setIsSeeMoreOpen(!isSeeMoreOpen)}>
					<p>{expandBtn.text}</p>
					<i className={`fa-solid ${expandBtn.icon}`}></i>
				</button>
			</div>
		</div>
	)
}

export default ProductAbout
