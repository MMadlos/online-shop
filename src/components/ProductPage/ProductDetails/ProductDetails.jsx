import AddCartBtn from "../../Elements/AddCartBtn"
import Rating from "./Rating"
import ProductAbout from "./ProductAbout"
import promoImage from "/public/promo-image-tiny.png"

function ProductDetails({ productSelected }) {
	const { id, name, price, url, rate, countRates, description } = productSelected

	return (
		<div
			className="product-container"
			data-product-id={id}>
			<img
				src={url}
				alt="#"
				className="product-image"
			/>

			<div className="product-details">
				<img
					src={promoImage}
					alt=""
					className="promo-image"
				/>
				<div className="product-header-container">
					<h1 className="product-header">{name}</h1>
					<Rating
						rate={rate}
						countRates={countRates}
					/>
					<p className="price">{`$ ${price}`}</p>
				</div>

				<AddCartBtn />
				<ProductAbout description={description} />
			</div>
		</div>
	)
}

export default ProductDetails
