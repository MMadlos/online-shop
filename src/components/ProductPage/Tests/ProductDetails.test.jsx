import { describe, it, expect, vi } from "vitest"
import { screen, render } from "@testing-library/react"

import ProductDetails from "../ProductDetails/ProductDetails"

vi.mock("../ProductDetails/ProductAbout", () => ({
	default: ({ description }) => {
		return <div data-testid="produt-about">{description}</div>
	},
}))

const mockProductSelected = {
	id: 1,
	name: "Product 1",
	price: "ProductPrice",
	url: "Url",
	rate: "ProductRate",
	countRates: "CountRates",
	description: "Description",
}

describe("ProductDetails", () => {
	it.skip("renders component", () => {
		const { container } = render(<ProductDetails productSelected={mockProductSelected} />)

		expect(container).toMatchSnapshot()
	})

	it("renders image and it contains product url", () => {
		render(<ProductDetails productSelected={mockProductSelected} />)

		const allIMG = screen.getAllByRole("img")

		expect(allIMG).toHaveLength(2)
		expect(allIMG[0]).toBeInTheDocument()
		expect(allIMG[0]).toHaveAttribute("src", mockProductSelected.url)
	})

	it("renders promoImage and its src is not empty", () => {
		render(<ProductDetails productSelected={mockProductSelected} />)

		const allIMG = screen.getAllByRole("img")

		expect(allIMG).toHaveLength(2)
		expect(allIMG[1]).toBeInTheDocument()
		expect(allIMG[1]).toHaveAttribute("src", expect.not.stringContaining(" "))
	})

	it("renders product price", () => {
		render(<ProductDetails productSelected={mockProductSelected} />)

		const regex = new RegExp(mockProductSelected.price)

		const price = screen.getByText(regex)
		expect(price).toBeInTheDocument()
	})

	it("renders rate and countRates", () => {
		render(<ProductDetails productSelected={mockProductSelected} />)

		const regexRates = new RegExp(mockProductSelected.rate)
		const regexCountRates = new RegExp(mockProductSelected.countRates)

		const rates = screen.getByText(regexRates)
		const countRates = screen.getByText(regexCountRates)

		expect(rates).toBeInTheDocument()
		expect(countRates).toBeInTheDocument()
	})

	it("renders description", () => {
		render(<ProductDetails productSelected={mockProductSelected} />)

		const regex = new RegExp(mockProductSelected.description)
		const description = screen.getByText(regex)

		expect(description).toBeInTheDocument()
	})
})
