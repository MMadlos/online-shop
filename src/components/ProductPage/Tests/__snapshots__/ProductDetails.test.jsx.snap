// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductDetails > renders component 1`] = `
<div>
  <div
    class="product-container"
    data-product-id="1"
  >
    <img
      alt="#"
      class="product-image"
      src="Url"
    />
    <div
      class="product-details"
    >
      <img
        alt=""
        class="promo-image"
        src="/src/assets/promo-image-tiny.png"
      />
      <div
        class="product-header-container"
      >
        <h1
          class="product-header"
        >
          Product 1
        </h1>
        <div
          class="rate-container"
        >
          <div
            class="star-container"
          >
            <i
              class="fa-solid fa-star"
            />
            <i
              class="fa-solid fa-star"
            />
          </div>
          <p
            class="rate"
          >
            3 (4)
          </p>
        </div>
        <p
          class="price"
        >
          $ 2
        </p>
      </div>
      <button
        id="add-cart"
      >
        Add to cart
      </button>
      <div
        class="product-about-container"
      >
        <p>
          About the product
        </p>
        <div
          class="product-about"
        >
          <p
            class="description"
          >
            Description
          </p>
          <button
            class="expand-btn none"
          >
            <p>
              See more
            </p>
            <i
              class="fa-solid fa-chevron-down"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
